README.md
# SocialsRegex for JavaScript

Detect and extract URLs of social media profiles with ease. This JavaScript package provides regular expressions and utilities to identify and extract social media account URLs from text.

## Installation

Install the package using npm:

```bash
npm install socials_regex
```

## Usage

You can read Full Docs here

- [Docs](https://github.com/talaatmagdyx/js_socials_regex/blob/master/docs/README.md)
- [Example](https://github.com/talaatmagdyx/js_socials_regex/blob/master/examples)

#### extractMatchesPerPlatform:

Witness the magic of extracting matches for each social media platform within the provided text.

```javascript
const {
  SocialsRegex,
  SocialExtraction,
  PlatformsRegex,
} = require("socials_regex");

// Example text containing social platform URLs
const text = `
  Check out my GitHub: https://github.com/example_user
  Also, find me on Twitter: https://twitter.com/example_twitter
   Also, find me on Twitter: https://twitter.com/example_twitter
  Also, find me on Twitter: https://twitter.com/example_twitter

  Contact me on LinkedIn: https://linkedin.com/in/example_linkedin
`;

// Use the static method directly
const matchesPerPlatform = SocialExtraction.extractMatchesPerPlatform(text);
console.log("Matches per platform:", matchesPerPlatform);
/* 
 * Matches per platform: {
  twitter: { user: [ [Object], [Object], [Object] ] },
  linkedin: { profile: [ [Object] ] },
  github: { user: [ [Object] ] }
}
* */
```

#### extractMatchesByPlatform:

Perform wizardry to extract matches based on regex patterns defined for a specific social media platform.

```javascript
const { SocialsRegex, SocialExtraction } = require("socials_regex");

// Example usage in your application
const text = "Visit my Yelp page: https://www.yelp.com/biz/example-business";
const platform = SocialsRegex.Platforms.PLATFORM_YELP;
const matches = SocialExtraction.extractMatchesByPlatform(platform, text);

console.log(matches[SocialsRegex.Platforms.PLATFORM_YELP].company);
// Output: [{ matched: 'https://www.yelp.com/biz/example-business', company: 'example-business' }]
```

another example

```javascript
const {
  SocialsRegex,
  SocialExtraction,
  PlatformsRegex,
} = require("socials_regex");

// Example text containing social platform URLs
const text = `
  Check out my GitHub: https://github.com/example_user
  Also, find me on Twitter: https://twitter.com/example_twitter
   Also, find me on Twitter: https://twitter.com/example_twitter
  Also, find me on Twitter: https://twitter.com/example_twitter

  Contact me on LinkedIn: https://linkedin.com/in/example_linkedin
`;

const twitterMatches = SocialExtraction.extractMatchesByPlatform(
  SocialsRegex.Platforms.PLATFORM_TWITTER,
  text,
);
console.log("Twitter matches:", twitterMatches);
// Twitter matches: { twitter: { user: [ [Object], [Object], [Object] ] } }
```

#### extractMatchesByRegex:

Cast a spell to extract matches based on the provided regex pattern within the text.

```javascript
const {
  SocialsRegex,
  SocialExtraction,
  PlatformsRegex,
} = require("socials_regex");

// Example text containing social platform URLs
const text = `
  Check out my GitHub: https://github.com/example_user
  Also, find me on Twitter: https://twitter.com/example_twitter
   Also, find me on Twitter: https://twitter.com/example_twitter
  Also, find me on Twitter: https://twitter.com/example_twitter

  Contact me on LinkedIn: https://linkedin.com/in/example_linkedin
`;

const githubMatches = SocialExtraction.extractMatchesByRegex(
  [
    PlatformsRegex.REGEX[SocialsRegex.Platforms.PLATFORM_GITHUB].user,
    PlatformsRegex.REGEX[SocialsRegex.Platforms.PLATFORM_TWITTER].user,
  ],
  text,
);
console.log("GitHub matches:", githubMatches);
/*
GitHub matches: [
  [* 

    {
      matched: 'https://github.com/example_user',
      login: 'example_user'
    }
  ],
  [
    {
      matched: 'https://twitter.com/example_twitter',
      username: 'example_twitter'
    },
    {
      matched: 'https://twitter.com/example_twitter',
      username: 'example_twitter'
    },
    {
      matched: 'https://twitter.com/example_twitter',
      username: 'example_twitter'
    }
  ]
]

* */
```

## Supported Platforms

```javascript
const {
  SocialsRegex,
  SocialExtraction,
  PlatformsRegex,
} = require("socials_regex");
const supportedPlatforms = PlatformsRegex.supportedPlatformsRegex();
console.log(supportedPlatforms)[
  // output
  ("yelp",
  "whatsapp",
  "stackexchange network",
  "crunchbase",
  "angellist",
  "xing",
  "vimeo",
  "telegram",
  "stackoverflow",
  "stackexchange",
  "snapchat",
  "skype",
  "reddit",
  "phone",
  "medium",
  "hackernews",
  "email",
  "youtube",
  "instagram",
  "twitter",
  "linkedin",
  "github",
  "facebook")
];

// OR

const {
  SocialsRegex,
  SocialExtraction,
  PlatformsRegex,
} = require("socials_regex");
const supportedPlatforms = SocialsRegex.Platforms.all();
console.log(supportedPlatforms)[
  ("PLATFORM_FACEBOOK",
  "PLATFORM_GITHUB",
  "PLATFORM_LINKEDIN",
  "PLATFORM_TWITTER",
  "PLATFORM_INSTAGRAM",
  "PLATFORM_YOUTUBE",
  "PLATFORM_EMAIL",
  "PLATFORM_HACKER_NEWS",
  "PLATFORM_MEDIUM",
  "PLATFORM_PHONE",
  "PLATFORM_REDDIT",
  "PLATFORM_SKYPE",
  "PLATFORM_SNAPCHAT",
  "PLATFORM_STACKEXCHANGE",
  "PLATFORM_STACKOVERFLOW",
  "PLATFORM_TELEGRAM",
  "PLATFORM_VIMEO",
  "PLATFORM_XING",
  "PLATFORM_ANGELLIST",
  "PLATFORM_CRUNCHBASE",
  "PLATFORM_STACKEXCHANGE_NETWORK",
  "PLATFORM_WHATSAPP",
  "PLATFORM_YELP",
  "all",
  "show")
];
```

## Supported Regexes

```javascript
const supportedRegexes = SocialsRegex.Regexes.all();
// Output: ['ANGELLIST_URL_REGEX', 'CRUNCHBASE_URL_REGEX', 'EMAIL_URL_REGEX', 'FACEBOOK_URL_REGEX', 'GITHUB_URL_REGEX', 'HACKERNEWS_URL_REGEX', ...]
```

## Features

- Detect the platform a URL points to (support for major platforms).
- Extract information contained within the URL without accessing the link.
- Extract emails and phone numbers from hyperlinks.

## Development

- Clone the repository: `git clone https://github.com/talaatmagdyx/js_socials_regex.git`
- Install dependencies: `npm install`
- Run tests: `npm test`

## References

- [social-media-profiles-regexs](https://github.com/lorey/social-media-profiles-regexs):
  extract urls of social media profiles with regular expressions
- [Ruby socials_regex](https://github.com/talaatmagdyx/socials_regex) Social Regex Account Detection and Extraction for Ruby. Detect and extract URLs of social accounts: throw in URLs, get back URLs of social media profiles by type.

## Contributing

Bug reports and pull requests are welcome on GitHub at [Contributing](https://github.com/talaatmagdyx/js_socials_regex/blob/master/.github/CONTRIBUTING.md). This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/talaatmagdyx/js_socials_regex/blob/master/.github/CODE_OF_CONDUCT.md).

## License

The package is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Reporting Bugs / Feature Requests

Please [open an issue on GitHub](https://github.com/talaatmagdyx/js_socials_regex/issues) for feedback, new feature requests, or bug reports.

## Pull Request

Please read [Contributing](https://github.com/talaatmagdyx/js_socials_regex/blob/master/.github/PULL_REQUEST_TEMPLATE.md)

## Code of Conduct

Everyone interacting in the SocialsRegex project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/talaatmagdyx/js_socials_regex/blob/master/.github/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [released]

## [1.0.0] - 2023-12-11

- first release

## [1.0.2] - 2023-12-12

- improve docs & deepsource

## [1.0.3] - 2023-12-14

- improve docs & more examples


test/test_regex.js
const { expect } = require('chai')
const SocialsRegex = require('../src/platforms')

describe(SocialsRegex.Platforms.PLATFORM_ANGELLIST, () => {
  it('Valid # company', () => {
    const inputStr = 'https://angel.co/company/twitter'
    const regex = SocialsRegex.Regexes.ANGELLIST_URL_REGEX.company
    expect(regex.test(inputStr)).to.be.true
  })

  it('Valid # job', () => {
    const inputStr =
      'https://angel.co/company/twitter/jobs/576275-engineering-manager'
    const regex = SocialsRegex.Regexes.ANGELLIST_URL_REGEX.job
    expect(regex.test(inputStr)).to.be.true
  })

  it('Valid # user ', () => {
    const inputStr = 'https://angel.co/p/naval'
    const anotherInputStr = 'https://angel.co/u/karllorey'
    const regex = SocialsRegex.Regexes.ANGELLIST_URL_REGEX.user
    expect(regex.test(inputStr)).to.be.true
    expect(regex.test(anotherInputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_CRUNCHBASE, () => {
  it('Valid # company', () => {
    const inputStr = 'http://crunchbase.com/organization/acme-corp'
    const regex = SocialsRegex.Regexes.CRUNCHBASE_URL_REGEX.company
    expect(regex.test(inputStr)).to.be.true
  })

  it('Valid # person', () => {
    const inputStr = 'http://crunchbase.com/person/karl-lorey'
    const regex = SocialsRegex.Regexes.CRUNCHBASE_URL_REGEX.person
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_EMAIL, () => {
  it('Valid # email', () => {
    const inputStr = 'jeff@amazon.com'
    const regex = SocialsRegex.Regexes.EMAIL_URL_REGEX.email
    expect(regex.test(inputStr)).to.be.true
  })

  it('Valid # email with mailto', () => {
    const inputStr = 'mailto:plususer+test@gmail.com'
    const regex = SocialsRegex.Regexes.EMAIL_URL_REGEX.email
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_FACEBOOK, () => {
  it('Valid # username with shortcut', () => {
    const inputStr = 'http://fb.com/peter_parker-miller'
    const regex = SocialsRegex.Regexes.FACEBOOK_URL_REGEX.username
    expect(regex.test(inputStr)).to.be.true
  })

  it('Valid # username', () => {
    const inputStr = 'https://facebook.com/peterparker'
    const regex = SocialsRegex.Regexes.FACEBOOK_URL_REGEX.username
    expect(regex.test(inputStr)).to.be.true
  })

  it('Valid # profile_id', () => {
    const inputStr = 'https://www.facebook.com/100004123456789'
    const regex = SocialsRegex.Regexes.FACEBOOK_URL_REGEX.profile_id
    expect(regex.test(inputStr)).to.be.true
  })

  it('Valid # profile_id with id params', () => {
    const inputStr = 'https://www.facebook.com/profile.php?id=100004123456789'
    const regex = SocialsRegex.Regexes.FACEBOOK_URL_REGEX.profile_id
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_GITHUB, () => {
  it('Valid # repo', () => {
    const inputStr = 'https://github.com/talaatmagdyx/socials_regex'
    const regex = SocialsRegex.Regexes.GITHUB_URL_REGEX.repo
    expect(regex.test(inputStr)).to.be.true
  })

  it('Valid # user', () => {
    const inputStr = 'https://github.com/talaatmagdyx'
    const regex = SocialsRegex.Regexes.GITHUB_URL_REGEX.user
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_HACKER_NEWS, () => {
  it('Valid # item', () => {
    const inputStr = 'https://news.ycombinator.com/item?id=23290375'
    const regex = SocialsRegex.Regexes.HACKERNEWS_URL_REGEX.item
    expect(regex.test(inputStr)).to.be.true
  })

  it('Valid # user', () => {
    const inputStr = 'https://news.ycombinator.com/user?id=CamelCaps'
    const regex = SocialsRegex.Regexes.HACKERNEWS_URL_REGEX.user
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_INSTAGRAM, () => {
  it('Valid # profile', () => {
    const inputStr = 'https://instagram.com/talaatmagdyx'
    const regex = SocialsRegex.Regexes.INSTAGRAM_URL_REGEX.profile
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_LINKEDIN, () => {
  it('Valid # company', () => {
    const inputStr =
      'https://fr.linkedin.com/school/université-grenoble-alpes/'
    const regex = SocialsRegex.Regexes.LINKEDIN_URL_REGEX.company
    expect(regex.test(inputStr)).to.be.true
  })

  it('Valid # post', () => {
    const inputStr =
      'https://www.linkedin.com/feed/update/urn:li:activity:6665508550111912345/'
    const regex = SocialsRegex.Regexes.LINKEDIN_URL_REGEX.post
    expect(regex.test(inputStr)).to.be.true
  })

  it('Valid # profile', () => {
    const inputStr = 'https://www.linkedin.com/in/talaatmagdyx/'
    const regex = SocialsRegex.Regexes.LINKEDIN_URL_REGEX.profile
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_MEDIUM, () => {
  it('Valid # post', () => {
    const inputStr = 'https://medium.com/does-exist/some-post-123abc'
    const regex = SocialsRegex.Regexes.MEDIUM_URL_REGEX.post
    expect(regex.test(inputStr)).to.be.true
  })

  it('Valid # post_of_subdomain_publication', () => {
    const inputStr =
      'https://onezero.medium.com/what-facebooks-remote-work-policy-means-for-the-future-of-tech-salaries-everywhere-edf859226b62?source=grid_home------'
    const regex =
      SocialsRegex.Regexes.MEDIUM_URL_REGEX.post_of_subdomain_publication
    expect(regex.test(inputStr)).to.be.true
  })

  it('Valid # user', () => {
    const inputStr = 'https://medium.com/@talaatmagdyx'
    const regex = SocialsRegex.Regexes.MEDIUM_URL_REGEX.user
    expect(regex.test(inputStr)).to.be.true
  })

  it('Valid # user_by_id', () => {
    const inputStr =
      'https://medium.com/u/b3d3d3653c2c?source=post_page-----da92b81b85ef----------------------'
    const regex = SocialsRegex.Regexes.MEDIUM_URL_REGEX.user_by_id
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_PHONE, () => {
  it('Valid # number', () => {
    const inputStr = 'tel:+49 900 123456'
    const regex = SocialsRegex.Regexes.PHONE_URL_REGEX.number
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_REDDIT, () => {
  it('Valid # user', () => {
    const inputStr = 'https://old.reddit.com/user/ar-guetita'
    const regex = SocialsRegex.Regexes.REDDIT_URL_REGEX.user
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_SKYPE, () => {
  it('Valid # profile', () => {
    const inputStr = 'skype:echo123?call'
    const regex = SocialsRegex.Regexes.SKYPE_URL_REGEX.profile
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_SNAPCHAT, () => {
  it('Valid # profile', () => {
    const inputStr = 'https://www.snapchat.com/add/example_user/'
    const regex = SocialsRegex.Regexes.SNAPCHAT_URL_REGEX.profile
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_STACKEXCHANGE, () => {
  it('Valid # user', () => {
    const inputStr = 'https://www.stackexchange.com/users/12345/example_user/'
    const regex = SocialsRegex.Regexes.STACKEXCHANGE_URL_REGEX.user
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_STACKEXCHANGE_NETWORK, () => {
  it('Valid # user', () => {
    const inputStr =
      'https://gaming.stackexchange.com/users/304007/talaat-magdy'
    const regex = SocialsRegex.Regexes.STACKEXCHANGE_NETWORK_URL_REGEX.user
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_STACKOVERFLOW, () => {
  it('Valid # question', () => {
    const inputStr = 'https://stackoverflow.com/questions/12345/how-to-embed'
    const regex = SocialsRegex.Regexes.STACKOVERFLOW_URL_REGEX.question
    expect(regex.test(inputStr)).to.be.true
  })

  it('Valid # user', () => {
    const inputStr = 'https://stackoverflow.com/users/13916928/talaat-magdy'
    const regex = SocialsRegex.Regexes.STACKOVERFLOW_URL_REGEX.user
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_TELEGRAM, () => {
  it('Valid # profile', () => {
    const inputStr = 'https://t.me/example_username/'
    const regex = SocialsRegex.Regexes.TELEGRAM_URL_REGEX.profile
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_TWITTER, () => {
  it('Valid # status', () => {
    const inputStr = 'https://twitter.com/karllorey/status/1259924082067374088'
    const regex = SocialsRegex.Regexes.TWITTER_URL_REGEX.status
    expect(regex.test(inputStr)).to.be.true
  })

  it('Valid # user', () => {
    const inputStr = 'https://twitter.com/talaatmagdyx'
    const regex = SocialsRegex.Regexes.TWITTER_URL_REGEX.user
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_VIMEO, () => {
  it('Valid # user', () => {
    const inputStr = 'https://vimeo.com/user12345'
    const regex = SocialsRegex.Regexes.VIMEO_URL_REGEX.user
    expect(regex.test(inputStr)).to.be.true
  })

  it('Valid # video', () => {
    const inputStr = 'https://vimeo.com/123456789'
    const regex = SocialsRegex.Regexes.VIMEO_URL_REGEX.video
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_XING, () => {
  it('Valid # profile', () => {
    const inputStr = 'https://www.xing.com/profile/Tobias_Zilbersahn5'
    const regex = SocialsRegex.Regexes.XING_URL_REGEX.profile
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_YOUTUBE, () => {
  it('Valid # channel', () => {
    const inputStr = 'https://www.youtube.com/channel/UCxyz123456789'
    const regex = SocialsRegex.Regexes.YOUTUBE_URL_REGEX.channel
    expect(regex.test(inputStr)).to.be.true
  })

  it('Valid # user', () => {
    const inputStr = 'https://www.youtube.com/user/username123'
    const regex = SocialsRegex.Regexes.YOUTUBE_URL_REGEX.user
    expect(regex.test(inputStr)).to.be.true
  })

  it('Valid # video', () => {
    const inputStr = 'https://www.youtube.com/watch?v=dQw4w9WgXcQ'
    const regex = SocialsRegex.Regexes.YOUTUBE_URL_REGEX.video
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_WHATSAPP, () => {
  it('Valid # phone', () => {
    const inputStr = 'https://wa.me/1234567890'
    const regex = SocialsRegex.Regexes.WHATSAPP_URL_REGEX.phone
    expect(regex.test(inputStr)).to.be.true
  })
})

describe(SocialsRegex.Platforms.PLATFORM_YELP, () => {
  it('Valid # company', () => {
    const inputStr = 'https://www.yelp.com/biz/example-business'
    const regex = SocialsRegex.Regexes.YELP_URL_REGEX.company
    expect(regex.test(inputStr)).to.be.true
  })
})


src/platforms.js
// Define the SocialsRegex module for handling social platform regex patterns
const SocialsRegex = {}

// Define the Platforms class in the module
SocialsRegex.Platforms = {
  PLATFORM_FACEBOOK: 'facebook',
  PLATFORM_GITHUB: 'github',
  PLATFORM_LINKEDIN: 'linkedin',
  PLATFORM_TWITTER: 'twitter',
  PLATFORM_INSTAGRAM: 'instagram',
  PLATFORM_YOUTUBE: 'youtube',
  PLATFORM_EMAIL: 'email',
  PLATFORM_HACKER_NEWS: 'hackernews',
  PLATFORM_MEDIUM: 'medium',
  PLATFORM_PHONE: 'phone',
  PLATFORM_REDDIT: 'reddit',
  PLATFORM_SKYPE: 'skype',
  PLATFORM_SNAPCHAT: 'snapchat',
  PLATFORM_STACKEXCHANGE: 'stackexchange',
  PLATFORM_STACKOVERFLOW: 'stackoverflow',
  PLATFORM_TELEGRAM: 'telegram',
  PLATFORM_VIMEO: 'vimeo',
  PLATFORM_XING: 'xing',
  PLATFORM_ANGELLIST: 'angellist',
  PLATFORM_CRUNCHBASE: 'crunchbase',
  PLATFORM_STACKEXCHANGE_NETWORK: 'stackexchange network',
  PLATFORM_WHATSAPP: 'whatsapp',
  PLATFORM_YELP: 'yelp',

  // Define the 'all' and 'show' methods in the Platforms class
  all () {
    return Object.keys(this)
  },

  show (const_name) {
    return this[const_name] ?? ''
  }
}

// Define the Regexes class in the module
SocialsRegex.Regexes = {
  ANGELLIST_URL_REGEX: {
    // https://angel.co/company/twitter, https://angel.co/company/twitter/culture
    company:
      /(?:https?:)?\/\/angel\.co\/company\/(?<company>[A-Za-z0-9_-]+)(?:\/(?<company_subpage>[A-Za-z0-9-]+))?/,
    // https://angel.co/company/twitter/jobs/576275-engineering-manager
    job: /(?:https?:)?\/\/angel\.co\/company\/(?<company>[A-Za-z0-9_-]+)\/jobs\/(?<job_permalink>(?<job_id>[0-9]+)-(?<job_slug>[A-Za-z0-9-]+))/,
    // https://angel.co/p/naval, https://angel.co/u/karllorey
    user: /(?:https?:)?\/\/angel\.co\/(?<type>u|p)\/(?<user>[A-Za-z0-9_-]+)/
  },

  CRUNCHBASE_URL_REGEX: {
    // http://crunchbase.com/organization/acme-corp
    company:
      /(?:https?:)?\/\/crunchbase\.com\/organization\/(?<organization>[A-Za-z0-9_-]+)/,
    // http://crunchbase.com/person/karl-lorey
    person:
      /(?:https?:)?\/\/crunchbase\.com\/person\/(?<person>[A-Za-z0-9_-]+)/
  },

  EMAIL_URL_REGEX: {
    // jeff@amazon.com, mailto:jeff@amazon.com, mailto:plususer+test@gmail.com
    email: /(?:mailto:)?(?<email>[A-Za-z0-9_.+-]+@[A-Za-z0-9_.-]+\.[A-Za-z]+)/,
    email_without_extract: /^[\w+\-.]+@[a-z\d-]+(\.[a-z\d-]+)*\.[a-z]+$/i
  },
  FACEBOOK_URL_REGEX: {
    // http://fb.com/peter_parker-miller, https://facebook.com/peter.parker, https://facebook.com/peterparker
    username:
      /(?:https?:)?\/\/(?:www\.)?(?:facebook|fb)\.com\/(?<profile>(?![A-Za-z]+\.php)(?!marketplace|gaming|watch|me|messages|help|search|groups)[A-Za-z0-9_\-.]+)\/?/,
    // https://www.facebook.com/100004123456789, https://www.facebook.com/profile.php?id=100004123456789
    profile_id:
      /(?:https?:)?\/\/(?:www\.)?facebook\.com\/(?:profile\.php\?id=)?(?<id>[0-9]+)/
  },
  GITHUB_URL_REGEX: {
    // https://github.talaatmagdyx/socials_regexcom/
    repo: /(?:https?:)?\/\/(?:www\.)?github\.com\/(?<login>[A-Za-z0-9_-]+)\/(?<repo>[A-Za-z0-9_-]+)\/?/,
    // https://github.com/talaatmagdyx
    user: /(?:https?:)?\/\/(?:www\.)?github\.com\/(?<login>[A-Za-z0-9_-]+)\/?/
  },
  HACKERNEWS_URL_REGEX: {
    // An item can be a post or a direct link to a comment.
    // https://news.ycombinator.com/item?id=23290375
    item: /(?:https?:)?\/\/news\.ycombinator\.com\/item\?id=(?<item>[0-9]+)/,
    // https://news.ycombinator.com/user?id=CamelCaps
    user: /(?:https?:)?\/\/news\.ycombinator\.com\/user\?id=(?<user>[A-Za-z0-9_-]+)/
  },
  INSTAGRAM_URL_REGEX: {
    // The rules:
    //  Matches with one . in them disco.dude but not two .. disco..dude
    //  Ending period not matched discodude.
    //  Match underscores _disco__dude
    //  Max characters of 30 1234567890123456789012345678901234567890
    //  https://instagram.com/__disco__dude
    profile:
      /(?:https?:)?\/\/(?:www\.)?(?:instagram\.com|instagr\.am)\/(?<username>[A-Za-z0-9_](?:(?:[A-Za-z0-9_]|(?:\.(?!\.))){0,28}(?:[A-Za-z0-9_]))?)/
  },
  LINKEDIN_URL_REGEX: {
    // https://fr.linkedin.com/school/université-grenoble-alpes/
    company:
      /(?:https?:)?\/\/(?:\w+\.)?linkedin\.com\/(?<company_type>(?:company|school))\/(?<company_permalink>[A-Za-z0-9\-À-ÿ.]+)\/?/,
    // https://www.linkedin.com/feed/update/urn:li:activity:6665508550111912345/
    post: /(?:https?:)?\/\/(?:\w+\.)?linkedin\.com\/feed\/update\/urn:li:activity:(?<activity_id>[0-9]+)\/?/,
    // https://www.linkedin.com/in/talaatmagdyx/
    profile:
      /(?:https?:)?\/\/(?:\w+\.)?linkedin\.com\/in\/(?<permalink>[\w\-_À-ÿ%]+)\/?/
  },
  MEDIUM_URL_REGEX: {
    // https://medium.com/does-exist/some-post-123abc
    post: /(?:https?:)?\/\/medium\.com\/(?:(?:@(?<username>[A-Za-z0-9]+))|(?<publication>[a-z-]+))\/(?<slug>[a-z0-9-]+)-(?<post_id>[A-Za-z0-9]+)(?:\?.*)?/,
    // https://onezero.medium.com/what-facebooks-remote-work-policy-means-for-the-future-of-tech-salaries-everywhere-edf859226b62?source=grid_home------
    post_of_subdomain_publication:
      /(?:https?:)?\/\/(?<publication>(?!www)[a-z-]+)\.medium\.com\/(?<slug>[a-z0-9-]+)-(?<post_id>[A-Za-z0-9]+)(?:\?.*)?/,
    // https://medium.com/@talaatmagdyx
    user: /(?:https?:)?\/\/medium\.com\/@(?<username>[A-Za-z0-9]+)(?:\?.*)?/,
    // Now redirects to new user profiles. Follow with a head or get request.
    // https://medium.com/u/b3d3d3653c2c?source=post_page-----da92b81b85ef----------------------
    user_by_id:
      /(?:https?:)?\/\/medium\.com\/u\/(?<user_id>[A-Za-z0-9]+)(?:\?.*)/
  },
  PHONE_URL_REGEX: {
    // Should be cleaned afterwards to strip dots, spaces, etc.
    // tel:+49 900 123456
    // tel:+49900123456
    number: /(?:tel|phone|mobile):(?<number>\+?[0-9. -]+)/
  },
  REDDIT_URL_REGEX: {
    // https://old.reddit.com/user/ar-guetita
    // https://reddit.com/u/ar-guetita
    user: /(?:https?:)?\/\/(?:[a-z]+\.)?reddit\.com\/(?:u(?:ser)?)\/(?<username>[A-Za-z0-9\-_]*)\/?/
  },
  SKYPE_URL_REGEX: {
    // Matches Skype's URLs to add contact, call, chat. More info at Skype SDK's docs.
    // Examples:
    // skype:echo123
    // skype:echo123?call
    profile:
      /(?:(?:callto|skype):)(?<username>[a-z][a-z0-9.,\-_]{5,31})(?:\?(?:add|call|chat|sendfile|userinfo))?/
  },
  SNAPCHAT_URL_REGEX: {
    // https://www.snapchat.com/add/example_user/
    profile:
      /(?:https?:)?\/\/(?:www\.)?snapchat\.com\/add\/(?<username>[A-Za-z0-9._-]+)\/?/
  },
  STACKEXCHANGE_URL_REGEX: {
    // This is the meta-platform above stackoverflow, etc. Username can be changed at any time, user_id is persistent.
    //  https://www.stackexchange.com/users/12345/example_user/
    user: /(?:https?:)?\/\/(?:www\.)?stackexchange\.com\/users\/(?<id>[0-9]+)\/(?<username>[A-Za-z0-9\-_.]+)\/?/
  },
  STACKEXCHANGE_NETWORK_URL_REGEX: {
    // While there are some "named" communities in the stackexchange network like stackoverflow,
    // many only exist as subdomains, i.e. gaming.stackexchange.com.
    // Again, username can be changed at any time, user_id is persistent.
    // https://gaming.stackexchange.com/users/304007/talaat-magdy
    user: /(?:https?:)?\/\/(?:(?<community>[a-z]+(?!www))\.)?stackexchange\.com\/users\/(?<id>[0-9]+)\/(?<username>[A-Za-z0-9\-_.]+)\/?/
  },
  STACKOVERFLOW_URL_REGEX: {
    // https://stackoverflow.com/questions/12345/how-to-embed
    question:
      /(?:https?:)?\/\/(?:www\.)?stackoverflow\.com\/questions\/(?<id>[0-9]+)\/(?<title>[A-Za-z0-9\-_.]+)\/?/,
    // Username can be changed at any time, user_id is persistent.
    // https://stackoverflow.com/users/13916928/talaat-magdy
    user: /(?:https?:)?\/\/(?:www\.)?stackoverflow\.com\/users\/(?<id>[0-9]+)\/(?<username>[A-Za-z0-9\-_.]+)\/?/
  },
  TELEGRAM_URL_REGEX: {
    // Matches for t.me, telegram.me and telegram.org.
    // Examples:
    // https://t.me/example_username/
    profile:
      /(?:https?:)?\/\/(?:t(?:elegram)?\.me|telegram\.org)\/(?<username>[a-z0-9_]{5,32})\/?/
  },
  TWITTER_URL_REGEX: {
    // https://twitter.com/karllorey/status/1259924082067374088
    status:
      /(?:https?:)?\/\/(?:[A-Za-z]+\.)?twitter\.com\/@?(?<username>[A-Za-z0-9_]+)\/status\/(?<tweet_id>[0-9]+)\/?/,
    // Allowed for usernames are alphanumeric characters and underscores.
    // http://twitter.com/@talaatmagdyx
    // http://twitter.com/talaatmagdyx
    // https://twitter.com/talaatmagdyx
    user: /(?:https?:)?\/\/(?:[A-Za-z]+\.)?twitter\.com\/@?(?!home|share|privacy|tos)(?<username>[A-Za-z0-9_]+)\/?/
  },
  VIMEO_URL_REGEX: {
    // https://vimeo.com/user12345
    user: /(?:https?:)?\/\/vimeo\.com\/user(?<id>[0-9]+)/,
    // https://vimeo.com/123456789
    // https://player.vimeo.com/video/148751763
    video:
      /(?:https?:)?\/\/(?:(?:www\.)?vimeo\.com|player\.vimeo\.com\/video)\/(?<id>[0-9]+)/
  },
  XING_URL_REGEX: {
    // Default slugs are Firstname_Lastname. If several people with the same name exist, a number is appended.
    // https://www.xing.com/profile/Tobias_Zilbersahn5
    profile:
      /(?:https?:)?\/\/(?:www\.)?xing\.com\/profile\/(?<slug>[A-Za-z0-9\-_]+)/
  },
  YOUTUBE_URL_REGEX: {
    // https://www.youtube.com/channel/UCxyz123456789
    channel:
      /(?:https?:)?\/\/(?:[A-Za-z]+\.)?youtube\.com\/channel\/(?<id>[A-Za-z0-9\-_]+)/,
    // https://www.youtube.com/user/username123
    user: /(?:https?:)?\/\/(?:[A-Za-z]+\.)?youtube\.com\/user\/(?<username>[A-Za-z0-9]+)/,
    // https://www.youtube.com/watch?v=dQw4w9WgXcQ
    // https://www.youtube.com/watch?v=dQw4w9WgXcQ
    // https://www.youtube.com/embed/dQw4w9WgXcQ
    // https://www.youtube.com/watch?v=6_b7RDuLwcI
    video:
      /(?:https?:)?\/\/(?:(?:www\.)?youtube\.com\/(?:watch\?v=|embed\/)|youtu\.be\/)(?<id>[A-Za-z0-9\-_]+)/
  },

  WHATSAPP_URL_REGEX: {
    // https://wa.me/1234567890
    phone: /(?:https?:)?\/\/(?:wa\.me\/)(?<number>\+?[0-9. -]+)/
  },

  YELP_URL_REGEX: {
    // https://www.yelp.com/biz/example-business
    company:
      /(?:https?:\/\/)?(?:www\.)?yelp\.com\/biz\/(?<company>[A-Za-z0-9_-]+)/
  },

  isMatch (input_str, regex) {
    return regex.test(input_str)
  },

  // Define the 'all' method in the Regexes class
  all () {
    return Object.keys(this)
  }
}

// Export the SocialsRegex module for use in other files
module.exports = SocialsRegex
